src/audio/module_adapter/module_adapter.c

```diff
@@ -1072,13 +1072,27 @@ static int module_adapter_copy_dp_queues(struct comp_dev *dev)
 	list_for_item(blist, &dev->bsink_list) {
 		/* output - we need to copy data from dp_queue (as source)
 		 * to audio_stream (as sink)
+		 *
+		 * a trick is needed there as a workaround
+		 * DP may produce a huge chunk of output data (i.e. 10 LL
+		 * cycles), and the following module should be able to consume it in 1 cycle chunks
+		 *
+		 * unfortunately some modules are not prepared to work when there's more than
+		 * 1 data portion available in the buffer and are draining buffers with data loss
```

Comment:
Isn't this configurable in topology, i.e. to make the downstream LL module consume 10 cycles ? @ranj063 did you mention to me there was a way to do this now ?

Reply:
it is not up to topology, it is up to each module separately
i.e. dc block - process all data from input, no matter how much data is there
```
static int dcblock_process(struct processing_module *mod,
			   struct input_stream_buffer *input_buffers,
			   int num_input_buffers,
			   struct output_stream_buffer *output_buffers,
			   int num_output_buffers)
{
	struct comp_data *cd = module_get_private_data(mod);
	struct audio_stream *source = input_buffers[0].data;
	struct audio_stream *sink = output_buffers[0].data;

	>>>>>>>   uint32_t frames = input_buffers[0].size;  <<<<<<

	comp_dbg(mod->dev, "dcblock_process()");

	cd->dcblock_func(cd, source, sink, frames);

	module_update_buffer_position(&input_buffers[0], &output_buffers[0], frames);
	return 0;
}```

process all data in buffer

Reply:
@singalsu @andrula-song fyi - lets make sure we check all modules to make sure they align with this use case.

Reply:
then maybe we have to use audio_stream_avail_frames instead of audio_stream_avail_frames_aligned for all components to get the exact data size to process so every component can process the same data size, or we force all components use the same alignment setting.

Reply:
@andrula-song do we need alignment at all? Especially for LL modules?

1) all buffers are aligned to cacheline size (64 bytes)
2) in LL processing a module is **always** getting a 1ms data portion from input and put 1ms to output  

that means - each buffer between 2 LLs should
 - be no larger than 1ms of data (max data rate in case of variable data rate)
 - at each LL cycle - be filled with 1ms AND drained 1ms (that means - IBS and OBS of connected modules should match)

so  
- always align to 64 bytes, 
- buffer always drained to empty
- always start from buffer begin (64bytes align)
- always linear data 

so what is the align useful for?  

real align is in IBSes and OBSes, and it is up to module to set in correctly - meaning that IBS/OBS should be aligned to frame size (in most cases bytes_per_sample*channels*rate_per_1ms*processing_period - i.e. 32bit/48000/2 = 384 for 1ms). And in LL a module is expected to consume/produce declared amount of bytes in each cycle. Any mismatch here means underrun/overrun (with small exception during pipeline startup)

Reply:
hi @marcinszkudlinski 

> in LL processing a module is always getting a 1ms data portion from input and put 1ms to output

 I recall once we got 47 or 49 frames(1 period is 48 frames) at the begining on tgl device, and if we use audio_stream_avail_frames_aligned(eg 8bytes and 1 frame aligned), there will always be 3 or 1 frames can not be consumed at this cycle and they will be processed in next cycle.
And if the components in the same pipeline have different alignment settings, there is the possibility that they can get different frames to process(eg: eq-fir algorithm asks 2 frames align while others ask 1).

Reply:
ok, so we can use aligned for now, but this corner case should be handled somehow

_off topic begin_
why: we're wasting a lot of memory for buffers between LL modules - a module for each bind, most of the time empty, At the moment there's enough memory to cover this, but nor for long
Required optimalization - use a single memory space as a reusable buffer, connected dynamically at in/out of module that is currently processing. 

As side-effect there won't be a place to keep a frame to be processed in next cycle
_off topic end_

Reply:
> i.e. dc block - process all data from input, no matter how much data is there

That's not correct to say. Unless module adapter has changed, the input_buffers[0].size in legacy audio stream mode used to contain a safe amount to process from source available and sink free, and also follow the defined data align constraints.  The struct field name "size" seems wrong though.

I've noticed this has changed in sink/source API we need a new way in modules to know the processing frames count.

Reply:
@marcinszkudlinski re offtopic - @btian1 is looking at this now, target the low hanging fruit 1st.

Reply:
@marcinszkudlinski @andrula-song @singalsu re -alignment - yes we need to always process on aligned SIMD sizes other wise
1) We need extra logic/MCPS to deal with unaligned leading data
2) We need extra logic/MCPS to deal with unaligned trailing data
3) additional impact of 1 & 2 is more complexity, larger TEXT, and decreased I$ perf.